soundcalc/zkvms/whir_based_vm.py

```diff
@@ -0,0 +1,470 @@
+
+from dataclasses import dataclass
+
+from soundcalc.common.fields import FieldParams, field_element_size_bits
+from soundcalc.common.utils import get_bits_of_security_from_error, get_size_of_merkle_path_bits
+from soundcalc.proxgaps.johnson_bound import JohnsonBoundRegime
+from soundcalc.proxgaps.proxgaps_regime import ProximityGapsRegime
+from soundcalc.proxgaps.unique_decoding import UniqueDecodingRegime
+from soundcalc.zkvms.zkvm import zkVM
+from typing import Tuple
+import math
+
+@dataclass(frozen=True)
+class WHIRBasedVMConfig:
+    """
+    A configuration of a FRI-based zkVM
+    """
+
+    # Name of the proof system
+    name: str
+
+    # The output length of the hash function that is used in bits
+    # Note: this concerns the hash function used for Merkle trees
+    hash_size_bits: int
+
+    # Parameters are inspired by Giacomo's script here for inspiration
+    # https://github.com/WizardOfMenlo/stir-whir-scripts/blob/main/src/whir.rs
+
+    # log2(1/rate), e.g., 2 if rate is 1/4
+    # note that this is the rate of the initial code
+    log_inv_rate: int
+
+    # this is the number of WHIR iterations
+    # note that a WHIR iteration consists of multiple rounds
+    # this is denoted by M in the paper
+    num_iterations: int
+
+    # this is what is called k_0,...,k_{M-1} in the paper
+    # as in Giacomo's script, we assume that this is the same for all
+    # see https://github.com/WizardOfMenlo/stir-whir-scripts/blob/main/src/whir.rs#L72
+    # in each iteration, we go from dimension 2^m_i to dimension 2^(m_i - k)
+    folding_factor: int
+
+    # the field that is used
+    field: FieldParams
+
+    # the log2 of the degree that we test
+    # in the WHIR paper, this is denoted by m
+    log_degree: int
+
+    # how many functions do we check in one go, i.e., this is the batch size
+    # for reference: https://github.com/WizardOfMenlo/stir-whir-scripts/blob/main/src/whir.rs#L144
+    batch_size: int
```

Comment:
Is this the same field as `num_polys` or `num_functions` for FRI-based VMs? Should we rename this field so that it matches the FRI configuration?